app/src/main/java/net/gsantner/markor/frontend/filesearch/FileSearchEngine.java

```diff
@@ -186,77 +196,78 @@ public void onDismissed(Snackbar snackbar, int event) {
         }
 
         @Override
-        protected List<FitFile> doInBackground(Void... voidp) {
-            Queue<File> mainQueue = new LinkedList<>();
-            mainQueue.add(_config.rootSearchDir);
-
-            while (!mainQueue.isEmpty() && !isCancelled() && !_isCanceled) {
-                File currentDirectory = mainQueue.remove();
-
-                if (!currentDirectory.canRead() || currentDirectory.isFile()) {
-                    continue;
-                }
-
-                _currentSearchDepth = getDirectoryDepth(_config.rootSearchDir, currentDirectory);
-                if (_currentSearchDepth > _config.maxSearchDepth) {
-                    break;
+        protected List<FitFile> doInBackground(final Void... ignored) {
+            final Stack<Pair<File, Integer>> stack = new Stack<>();
+            stack.add(Pair.create(_config.rootSearchDir, 0));
+            final int trimLength = _config.rootSearchDir.getAbsolutePath().length() + 1;
+
+            while (!stack.isEmpty() && !isCancelled() && !_isCanceled) {
+                final Pair<File, Integer> p = stack.pop();
+                final int depth = p.second;
+                final File dir = p.first;
+
+                if (depth < _config.maxSearchDepth && dir.canRead()) {
+                    handleDirectory(dir, trimLength, depth, stack);
+                    publishProgress(stack.size(), depth, _result.size(), _countCheckedFiles);
```

Comment:
Primary speedups:

1. Publish progress a little less frequently
2. Switch to stack vs queue, and simplify nesting
3. Only get canonical file when absolutely required (this is a huge bottleneck on android)